### PR TITLE
Hp tuning

### PR DESCRIPTION
**Pull request by NikkOLab**
MLDatasetTemplate
- Create and save log files for each training.
- User can select Loss function and Optimizer by setting its name as one of the parameters in the Create Training Request API. (user can check loss function and optimizer name in Keras official doc)
- Support hyperparameter tuning for "optimizer" and "learning_rate". The training will return the best model and save the best hyperparameters into the database.
- Currently not support other configs on Optimizer

### CODE REVIEW

File: training/MNISTFashionMicroservice/src/server/training.py [hunk 1]
```diff
@@ -2,20 +2,211 @@
...
+       self.linebuf = ''
+
+    def write(self, buf):
+       for line in buf.rstrip().splitlines():
+          self.logger.log(self.level, line.rstrip())
+
+    def flush(self):
+        pass
+
+# def train_model(training_id, model_data: ModelData):
```
Comment (jagath-jaikumar) [inline training/MNISTFashionMicroservice/src/server/training.py]: feel free to remove commented code

**Review by jagath-jaikumar — COMMENTED** [review]
Looks pretty close! Could you remove the commented out code and provide a simple test case or usage example in the readme so that we know how to try this out?

File: training/MNISTFashionMicroservice/src/server/training.py (hunk 1 above)
Comment (NikkOLab) [reply training/MNISTFashionMicroservice/src/server/training.py]: Ok, thanks! I will remove the commented code out and make another pull request. 

For the example/test case, I updated the Postman collection URL in the README to a new one which has an example in "Create Training Request" API.

Comment (jagath-jaikumar) [reply training/MNISTFashionMicroservice/src/server/training.py]: I think you can just update this pull request and I should be able to see the changes

Comment (NikkOLab) [reply training/MNISTFashionMicroservice/src/server/training.py]: okay! got it.

Comment (NikkOLab) [reply training/MNISTFashionMicroservice/src/server/training.py]: I updated the README file and removed the commented code out. Thanks

File: training/MNISTFashionMicroservice/src/server/training.py [hunk 2]
```diff
@@ -2,20 +2,68 @@
...
+    # SET LOGGER TO PRINT TO STDOUT AND WRITE TO FILE
+    logging.basicConfig(
+        level=logging.DEBUG,
+        format="%(asctime)s [%(levelname)s] %(message)s",
+        handlers=[
+            logging.FileHandler("/log/{}.log".format(training_id)),
+            logging.StreamHandler(sys.stdout)
+        ]
+    )
+    log = logging.getLogger('foobar')
```
Comment (CodyRichter) [inline training/MNISTFashionMicroservice/src/server/training.py]: What is this `foobar` logger? Is there a better name to give it?

Comment (jagath-jaikumar) [reply training/MNISTFashionMicroservice/src/server/training.py]: i think this a placeholder logger for when we implement one. that can be a separate MR

Comment (NikkOLab) [reply training/MNISTFashionMicroservice/src/server/training.py]: > What is this `foobar` logger? Is there a better name to give it?

Sure, I will change the name.   


> i think this a placeholder logger for when we implement one. that can be a separate MR

Thanks, what does MR mean?

Comment (NikkOLab) [reply training/MNISTFashionMicroservice/src/server/training.py]: Pushed. The logger name is changed.
